File: functions/src/lib/care-frequency.test.ts
import { describe, expect, it } from 'vitest';
import { CARE_WINDOW_DAYS, isExpected } from './care-frequency';

const now = new Date('2026-09-21T12:00:00Z'); // 14h à Paris
const daily = { timesPerDay: 1, everyDays: 1, enabled: true };
const everyTwoDays = { timesPerDay: 1, everyDays: 2, enabled: true };

describe('isExpected', () => {
  it('désactivé : jamais attendu', () => {
    expect(isExpected({ ...daily, enabled: false }, null, now)).toBe(false);
  });

  it("quotidien : toujours attendu, même fait aujourd'hui", () => {
    expect(isExpected(daily, null, now)).toBe(true);
    expect(isExpected(daily, new Date('2026-09-21T06:00:00Z'), now)).toBe(true);
  });

  it('tous les 2 jours : attendu sans historique ou après 2 jours civils', () => {
    expect(isExpected(everyTwoDays, null, now)).toBe(true);
    expect(isExpected(everyTwoDays, new Date('2026-09-20T16:00:00Z'), now)).toBe(false);
    expect(isExpected(everyTwoDays, new Date('2026-09-19T16:00:00Z'), now)).toBe(true);
  });

  it('jours civils de Paris : 23h30 UTC la veille est déjà « hier »', () => {
    // 2026-09-19T23:30Z = 20 septembre 01h30 à Paris → un seul jour civil avant le 21.
    expect(isExpected(everyTwoDays, new Date('2026-09-19T23:30:00Z'), now)).toBe(false);
  });

  it('la fenêtre couvre 7 jours', () => {
    expect(CARE_WINDOW_DAYS).toBe(7);
  });
});

File: test/features/baby/domain/care_frequency_test.dart
import 'package:colette/features/baby/domain/entities/care_frequency.dart';
import 'package:flutter_test/flutter_test.dart';

void main() {
  const daily = CareFrequency();
  final now = DateTime(2026, 9, 21, 14);

  group('previous (bouton −)', () {
    test('3/jour → 2/jour', () {
      expect(
        const CareFrequency(timesPerDay: 3).previous(),
        const CareFrequency(timesPerDay: 2),
      );
    });

    test('1/jour → tous les 2 jours', () {
      expect(daily.previous(), const CareFrequency(everyDays: 2));
    });

    test('tous les 2 jours → tous les 3 jours', () {
      expect(
        const CareFrequency(everyDays: 2).previous(),
        const CareFrequency(everyDays: 3),
      );
    });

    test('tous les 7 jours : butée', () {
      expect(const CareFrequency(everyDays: 7).previous(), isNull);
    });

    test('conserve enabled', () {
      expect(
        const CareFrequency(enabled: false).previous(),
        const CareFrequency(everyDays: 2, enabled: false),
      );
    });
  });

  group('next (bouton +)', () {
    test('tous les 3 jours → tous les 2 jours', () {
      expect(
        const CareFrequency(everyDays: 3).next(4),
        const CareFrequency(everyDays: 2),
      );
    });

    test('tous les 2 jours → 1/jour', () {
      expect(const CareFrequency(everyDays: 2).next(4), daily);
    });

    test('1/jour → 2/jour', () {
      expect(daily.next(4), const CareFrequency(timesPerDay: 2));
    });

    test('butée à maxTimesPerDay', () {
      expect(const CareFrequency(timesPerDay: 3).next(3), isNull);
    });
  });

  group('isExpected', () {
    test('désactivé : jamais attendu', () {
      expect(
        const CareFrequency(enabled: false)
            .isExpected(lastDoneAt: null, now: now),
        isFalse,
      );
    });

    test('1/jour : toujours attendu, même fait aujourd\'hui', () {
      expect(daily.isExpected(lastDoneAt: null, now: now), isTrue);
      expect(
        daily.isExpected(lastDoneAt: DateTime(2026, 9, 21, 8), now: now),
        isTrue,
      );
    });

    test('tous les 2 jours : attendu sans historique', () {
      expect(
        const CareFrequency(everyDays: 2)
            .isExpected(lastDoneAt: null, now: now),
        isTrue,
      );
    });

    test('tous les 2 jours, fait hier : pas attendu', () {
      expect(
        const CareFrequency(everyDays: 2)
            .isExpected(lastDoneAt: DateTime(2026, 9, 20, 18), now: now),
        isFalse,
      );
    });

    test('tous les 2 jours, fait avant-hier : attendu', () {
      expect(
        const CareFrequency(everyDays: 2)
            .isExpected(lastDoneAt: DateTime(2026, 9, 19, 18), now: now),
        isTrue,
      );
    });

    test('tous les 7 jours : pas attendu après 6 jours, attendu après 7', () {
      const weekly = CareFrequency(everyDays: 7);
      expect(
        weekly.isExpected(lastDoneAt: DateTime(2026, 9, 15, 18), now: now),
        isFalse,
      );
      expect(
        weekly.isExpected(lastDoneAt: DateTime(2026, 9, 14, 18), now: now),
        isTrue,
      );
    });

    test('DST : deux jours civils malgré le changement d\'heure', () {
      // Passage à l'heure d'été le 29 mars 2026 entre les deux instants.
      expect(
        const CareFrequency(everyDays: 2).isExpected(
          lastDoneAt: DateTime(2026, 3, 28, 20),
          now: DateTime(2026, 3, 30, 8),
        ),
        isTrue,
      );
    });
  });
}
